cutter: Initial implementation

ekimekim opened the pull request:
This series of commits introduces the cutter component.

The cutter has two jobs:
* To cut videos, taking them through states EDITED -> TRANSCODING
* To monitor TRANSCODING videos for when they're complete

This initial implementation only implements uploading to youtube,
and always to the same account. Later, we may want to implement alternate "upload locations"
which could be alternate youtube accounts, or other video services.

It could also use a lot more instrumentation.

NOTE: I have attempted to break this very large PR up into a series of commits that
introduces one change at a time. It's highly recommended you review by looking at the
commits one by one.

chrusher commented on cutter/cutter/youtube.py:
```diff
@@ -0,0 +1,107 @@
+		self._first_get_access_token.join()
+		return self._access_token
+
+	def get_access_token(self):
```
Doc strings please

chrusher commented on cutter/cutter/youtube.py:
```diff
@@ -0,0 +1,107 @@
+					self.logger.warning("Access token expires in {}s, less than normal leeway time of {}s".format(
+						expires_in, self.ACCESS_TOKEN_REFRESH_TIME_BEFORE_EXPIRY,
+					))
+				gevent.spawn_later(expires_in, self.get_access_token)
```
Does a stop in the parent Cutter or TranscodeChecker propagate throw to gevent.spawn_later()?

I might be missing something but the this seems to be called exactly at the expiry time and not the ACCESS_TOKEN_REFRESH_TIME_BEFORE_EXPIRY before the expiry.

ekimekim replied:
It does not. However, the extra work is harmless if it does happen, and in any case it almost certainly won't since you're generally only stopping when you're bringing down the entire process. The pending greenlet will not prevent process shutdown.

You're correct, good catch. Fixing.

ekimekim commented:
Addressed comments.
Currently there's no fancy backoffs in here, I don't think they're worth the complexity. They are at least random jittered, but tbh I don't think I'd want these retries any faster OR any slower - faster would be spammy and slower would be slow to recover.
